sql: TestDropWhileBackfill failed

sql.TestDropWhileBackfill [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4582385&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4582385&tab=artifacts#/) on master @ [be54b854824dc00dea8e3cf6289a1b020621c491](https://github.com/cockroachdb/cockroach/commits/be54b854824dc00dea8e3cf6289a1b020621c491):


```
=== RUN   TestDropWhileBackfill
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDropWhileBackfill3173526640
    test_log_scope.go:80: use -show-logs to present logs inline
    schema_changer_test.go:816: expected 402 key value pairs, but got 442
    panic.go:661: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDropWhileBackfill3173526640
--- FAIL: TestDropWhileBackfill (15.09s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/sql-schema
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDropWhileBackfill.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-13808

sql.TestDropWhileBackfill [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4587786&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4587786&tab=artifacts#/) on master @ [0184736b58da548c1ca476949b14edd0dbfa5a69](https://github.com/cockroachdb/cockroach/commits/0184736b58da548c1ca476949b14edd0dbfa5a69):


```
=== RUN   TestDropWhileBackfill
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDropWhileBackfill609874986
    test_log_scope.go:80: use -show-logs to present logs inline
    schema_changer_test.go:816: expected 402 key value pairs, but got 443
    panic.go:661: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDropWhileBackfill609874986
--- FAIL: TestDropWhileBackfill (13.99s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDropWhileBackfill.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql.TestDropWhileBackfill [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4585210&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4585210&tab=artifacts#/) on master @ [0184736b58da548c1ca476949b14edd0dbfa5a69](https://github.com/cockroachdb/cockroach/commits/0184736b58da548c1ca476949b14edd0dbfa5a69):


```
=== RUN   TestDropWhileBackfill
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/233e6aa31ff7ad29c896f0479cb5d0c4/logTestDropWhileBackfill624462638
    test_log_scope.go:80: use -show-logs to present logs inline
    schema_changer_test.go:816: expected 402 key value pairs, but got 443
    panic.go:661: -- test log scope end --
--- FAIL: TestDropWhileBackfill (11.63s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDropWhileBackfill.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql.TestDropWhileBackfill [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4592696&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4592696&tab=artifacts#/) on master @ [4cbaaaecd743adada69166216c455168d4345f84](https://github.com/cockroachdb/cockroach/commits/4cbaaaecd743adada69166216c455168d4345f84):


```
=== RUN   TestDropWhileBackfill
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDropWhileBackfill3759292425
    test_log_scope.go:80: use -show-logs to present logs inline
    schema_changer_test.go:816: expected 402 key value pairs, but got 442
    panic.go:661: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDropWhileBackfill3759292425
--- FAIL: TestDropWhileBackfill (15.01s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDropWhileBackfill.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql.TestDropWhileBackfill [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4596216&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4596216&tab=artifacts#/) on master @ [e9123b644529f1ab3afacc32b938e33707d87a02](https://github.com/cockroachdb/cockroach/commits/e9123b644529f1ab3afacc32b938e33707d87a02):


```
=== RUN   TestDropWhileBackfill
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/233e6aa31ff7ad29c896f0479cb5d0c4/logTestDropWhileBackfill477091332
    test_log_scope.go:80: use -show-logs to present logs inline
    schema_changer_test.go:816: expected 402 key value pairs, but got 442
    panic.go:661: -- test log scope end --
--- FAIL: TestDropWhileBackfill (12.61s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #78011 sql: TestDropWhileBackfill failed [C-test-failure O-robot T-sql-schema branch-release-22.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDropWhileBackfill.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql.TestDropWhileBackfill [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4604497&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4604497&tab=artifacts#/) on master @ [b407d3f19acba3a6159e23f78357a3e5585f6f4d](https://github.com/cockroachdb/cockroach/commits/b407d3f19acba3a6159e23f78357a3e5585f6f4d):


```
=== RUN   TestDropWhileBackfill
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDropWhileBackfill3431435325
    test_log_scope.go:80: use -show-logs to present logs inline
    schema_changer_test.go:816: expected 402 key value pairs, but got 443
    panic.go:661: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestDropWhileBackfill3431435325
--- FAIL: TestDropWhileBackfill (13.65s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #78011 sql: TestDropWhileBackfill failed [C-test-failure O-robot T-sql-schema branch-release-22.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDropWhileBackfill.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql.TestDropWhileBackfill [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4607114&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4607114&tab=artifacts#/) on master @ [9850f2aed2d44e04e08d9f14ff0c98a2bd193ee9](https://github.com/cockroachdb/cockroach/commits/9850f2aed2d44e04e08d9f14ff0c98a2bd193ee9):


```
=== RUN   TestDropWhileBackfill
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/233e6aa31ff7ad29c896f0479cb5d0c4/logTestDropWhileBackfill3125746347
    test_log_scope.go:80: use -show-logs to present logs inline
    schema_changer_test.go:816: expected 402 key value pairs, but got 443
    panic.go:661: -- test log scope end --
--- FAIL: TestDropWhileBackfill (12.67s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #78011 sql: TestDropWhileBackfill failed [C-test-failure O-robot T-sql-schema branch-release-22.1]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDropWhileBackfill.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -728,12 +728,12 @@ func TestDropWhileBackfill(t *testing.T) {
 			RunBeforeBackfillChunk: func(sp roachpb.Span) error {
 				if partialBackfillDone.Load().(bool) {
 					notifyBackfill()
-					// Returning DeadlineExceeded will result in the
-					// schema change being retried.
-					return context.DeadlineExceeded
 				}
 				partialBackfillDone.Store(true)
-				return nil
+				// Returning DeadlineExceeded will result in the
+				// schema change being retried and no data will be written
+				// to the new index.
+				return context.DeadlineExceeded
 			},
 		},
 		// Disable backfill migrations, we still need the jobs table migration.

